src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,182 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+// This file describes AST nodes for authorization logic
+
+#ifndef SRC_IR_AUTH_LOGIC_AST_H_
+#define SRC_IR_AUTH_LOGIC_AST_H_
+
+#include <string>
+#include <vector>
+#include <variant>
+
+namespace raksha::ir::auth_logic {
+
+class AstPrincipal {
+  public:
+    AstPrincipal(std::string name) : name_(std::move(name)) {}
+  private:
+    std::string name_;
+};
+
+// Used to represent whether a predicate is negated or not
+enum AstSign { negated, positive };
```

Comment:
Currently, these are not following the [style guide's decrees](https://google.github.io/styleguide/cppguide.html#Enumerator_Names) for naming enums.

In addition, you should consider [using `enum class` instead of `enum`](https://en.cppreference.com/w/cpp/language/enum). In C++, the classical behavior of `enum`s was that the constants declared therein would be added as members in the enclosing class or namespace. This is super weird to reason about, and so they added `enum class` in C++11, which causes the enum's type name to itself act as a name scope, to which the members belong.

Reply:
Also, move this into `AstPredicate` if it is only related to that class.

Reply:
- Fixed style
- Looks like I can't make Sign an inner class because it is used as a parameter for the constructor of Predicate